### PR TITLE
Add reasoning-effort thinking cutoff maps

### PR DESCRIPTION
**Pull request by ropoctl**
Lets an operator set a different native thinking cutoff per `reasoning_effort`
level, instead of one cutoff for every request.

`thinking_token_budget` now accepts a map alongside the legacy scalar:

```bash
--override-generation-config '{
  "thinking_token_budget": {"low": 1024, "medium": 4096, "xhigh": 8192}
}'
```

### Precedence

1. An explicit numeric `thinking_token_budget` on the request wins.
2. `-1` on the request opts out of the server cutoff entirely.
3. Otherwise the map is indexed by the request's `reasoning_effort`, falling
   back to `medium` when the request omits it or the map has no entry for it.
   Every map must therefore define `medium`; a map with unsupported keys, or
   without `medium`, is rejected at config time rather than at request time.
4. A scalar server value applies to every effort level, as before.

`reasoning_effort: none` disables thinking at the template layer, so it takes
no cutoff rather than the `medium` fallback.

### `max_tokens` semantics

`max_tokens` remains the client's total completion ceiling. The effective
cutoff is `min(selected_budget, max_tokens - 1)` — the one-token margin exists
only to inject the native thinking-end marker. There is no fixed final-answer
reserve; answer capacity is `max_tokens` minus the reasoning tokens actually
emitted, so clients wanting longer answers raise `max_tokens`.

### Enforcement fixes

Making per-effort cutoffs trustworthy required correcting four bugs in
`ThinkingBudgetStateHolder`, each covered by a test:

- **Row swap duplicated state.** Swapping a budgeted row with a plain row
  copied the state into both slots instead of moving it, so an unrelated
  request inherited a cutoff.
- **Prompt tokens charged against the cap.** The budget counts *generated*
  reasoning tokens, but a prompt that already opened a reasoning block had its
  prompt tokens deducted. Prompt tokens now only establish whether generation
  begins inside a reasoning block.
- **Markers split across chunks.** The incremental scan for multi-token think
  markers could miss one straddling two output chunks. The scan window now
  overlaps by `len(marker) - 1`.
- **Speculative overrun.** Under speculative decoding a proposed block could
  run past the cutoff before being accepted. The cutoff is now applied inside
  the speculated block, unless the draft closes the reasoning block naturally
  within the remaining budget.

### Docs

New `docs/reasoning-token-budgets.md` covers the schema, precedence, and a
Qwen3.8 example; README links to it from the existing thinking-budget section.

Scope: 8 files, +665/-25 — `vllm/sampling_params.py`,
`vllm/v1/sample/thinking_budget_state.py`, chat protocol, config, docs, tests.

## Summary by CodeRabbit

- **New Features**
  - Added configurable thinking-token budgets for reasoning models.
  - Supports effort-level budgets, request overrides, opt-out behavior, and template-level defaults.
  - Enforces budgets during generation and transitions to the final response when limits are reached.
  - Supports speculative decoding when compatible single-token reasoning markers are used.

- **Documentation**
  - Documented configuration, selection rules, limits, `max_tokens` behavior, and compatibility requirements.

- **Bug Fixes**
  - Improved reasoning-token counting, stream transitions, state reuse, and speculative decoding behavior.

### CODE REVIEW

**Comment by ehartford** [comment]
Pushed V2 model-runner support for thinking_token_budget to this branch (6741f1f9d, on top of a merge with current main).

Context: on the canonical SlimServe profiles — which all run the V2 runner — this feature previously 400'd per request, and the need is acute: an unbudgeted request on qwen38-nvfp4-1 spent all 1,500 completion tokens inside <think> and produced no answer.

Design: GPU-resident per-request state (remaining budget, in-think flag) enforced at the same in-place logits seam as the grammar bitmask, before the sampler/rejection-sampler dispatch — one mask serves both paths, and under speculative decoding a masked verify row rejects any draft token that is not the reasoning-end marker, while a draft that closes the block naturally within budget is left to rejection sampling (no special case needed). Draft consumption accumulates per request across the verify span using host-known geometry: no device syncs, no data-dependent shapes. Semantics match this PR's fixed V1 holder (generated in-think tokens only, markers uncharged, prompt-open starts at zero).

Scope: single-token reasoning markers (the qwen3 family); multi-token markers keep a per-request rejection that names the constraint and the V1 fallback. Nine unit tests plus a live gate on the canonical profile with spec decoding active: budget 64 → exactly 64 reasoning tokens then a complete answer; budget 16 → exactly 16; both finish=stop. Docs updated with a model-runner support section.

Two review items from earlier remain open for this branch: map validation still happens per request (the PR text says config time — either add boot-time validation or amend the text), and a template-level default reasoning_effort (SlimServe's Qwen3.8 profiles default to low via chat-template kwargs) does not inform map level selection, so such requests resolve to the medium entry.

File: vllm/v1/worker/gpu/sample/thinking_budget.py [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+        in_think0 = False
+        if prompt_token_ids:
+            arr = np.asarray(prompt_token_ids)
+            starts = np.flatnonzero(arr == self.start_id)
+            ends = np.flatnonzero(arr == self.end_id)
+            last_start = int(starts[-1]) if starts.size else -1
+            last_end = int(ends[-1]) if ends.size else -1
+            in_think0 = last_start > last_end
+        self.budget_np[req_idx] = budget
+        self._staged.append((req_idx, budget, in_think0, True))
```
[the remark this answers is not in the document]
Comment (ehartford) [reply vllm/v1/worker/gpu/sample/thinking_budget.py:108]: Verified and fixed in 9bd345585 — a real bug: the chat protocol resolves -1 to None, but a directly built SamplingParams (offline LLM API) carries -1 into add_request, which staged it as an ACTIVE budget of -1 and forced the end marker on the first in-think token. budget < 0 now stages inactive, with a mixed -1/positive regression test.

File: vllm/sampling_params.py [hunk 2]
```diff
@@ -55,11 +63,95 @@ def validate_thinking_token_budget(value: int | float | bool | None) -> int | No
...
+    ``reasoning_effort``. A map must contain ``medium``, which is both the
+    fallback for omitted request levels and for levels absent from a partial
+    map. An explicit request budget (including ``-1``) takes precedence.
+    """
+    configured_budget = request_budget
+    # An explicit API -1 must disable an operator-configured server default.
+    if configured_budget == -1:
+        return None
+    if configured_budget is None:
+        configured_budget = default_sampling_params.get("thinking_token_budget")
```
[the remark this answers is not in the document]
Comment (ehartford) [reply vllm/sampling_params.py:148]: Fixed in 9bd345585: the none check is hoisted above all budget resolution, so a scalar server default or explicit request budget no longer makes a no-thinking request look budgeted (which would needlessly trip the V2 multi-token-marker rejection). Coverage added for both cases.

File: vllm/config/vllm.py [hunk 3]
```diff
@@ -2318,11 +2318,17 @@ def _validate_v2_model_runner(self) -> None:
...
+            starts = self.reasoning_config.reasoning_start_token_ids or []
+            ends = self.reasoning_config.reasoning_end_token_ids or []
+            if len(starts) != 1 or len(ends) != 1:
+                logger.warning_once(
+                    "Model Runner V2 enforces thinking_token_budget only for "
+                    "single-token reasoning markers; this model's markers "
+                    "tokenize to multiple tokens, so the parameter will be "
+                    "rejected per request. Set VLLM_USE_V2_MODEL_RUNNER=0 if "
+                    "it is required."
+                )
```
[the remark this answers is not in the document]
Comment (ehartford) [reply vllm/config/vllm.py]: Confirmed and fixed in 9bd345585: _validate_v2_model_runner() runs at line 1551 and initialize_token_ids() at 1648, so the marker check could never fire. It now runs immediately after token-ID initialization, gated on use_v2_model_runner.

File: perf/optimization_status.md [hunk 4]
```diff
@@ -16313,3 +16313,47 @@ Load test (no-spec, in-process LLM, max_model_len 8192) iterated through:
...
+- Live gate (canonical qwen38-nvfp4-1, V2 + DFlash2 + throttle, seeded):
+  budget 64 -> exactly 64 reasoning tokens then a complete proof
+  (finish=stop); budget 16 -> exactly 16; no budget -> the pathological
+  all-thinking completion above. Suites: 167 passed / 105 skipped;
+  mypy clean across all touched files (incl. PR #13's pre-existing
+  debt).
+- Open from the PR review, left for the author or follow-up: map
+  validation still occurs per request (the PR text claims config time),
+  and template-level default reasoning_effort (our profiles' low) does
+  not inform map level selection.
```
[the remark this answers is not in the document]
Comment (ehartford) [reply perf/optimization_status.md:16373]: Amended in 9bd345585: the entry now carries baseline (the V2 400 + the all-thinking 1500-token run, with raw logs), hypothesis (logits-seam forcing per the grammar-bitmask precedent), correctness (9 unit tests + the token-exact live gate), throughput (inert path is one host-side numpy check per step; seeded bench heads byte-identical), decision (RETAINED), and raw artifact paths.

**Comment by ehartford** [comment]
The last open review item is closed: `f4a0da22b` moves the `thinking_token_budget` scalar/map validation to config time. `ModelConfig.get_diff_sampling_param` — which every serving handler constructor runs at startup — now calls the extracted `validate_thinking_token_budget_config`, so a malformed operator default (unknown map keys, a map without `medium`, invalid values) fails the server at boot instead of rejecting every request. The per-request path re-validates through the same helper for programmatically supplied maps.

Live-verified on the qwen38-nvfp4-1 checkpoint (dummy load): `{"thinking_token_budget": {"low": 100}}` exits 1 during serving init with the medium-required error and never reaches startup-complete; the control boot with `{"low": 1024, "medium": 4096}` reaches `Application startup complete`. 10 new config-time tests (5 bad shapes raise from `get_diff_sampling_param`, 5 good shapes pass through); both budget suites 52/52.

The PR description's "rejected at config time rather than at request time" now holds as written — no text amendment needed.
